src/Feldspar/Compiler/Imperative/Frontend.hs

```diff
@@ -50,62 +50,70 @@ type Location = Maybe (Expression ())
 
 -- | Copies expressions into a destination. If the destination is
 -- an array the arguments are appended into the destination.
-copyProg :: Location -> [Expression ()] -> Program ()
-copyProg _ []      = error "copyProg: missing source parameter."
-copyProg Nothing _ = Empty
-copyProg (Just outExp) inExp =
+copyProg :: Options -> Location -> [Expression ()] -> [Program ()]
+copyProg opts _ []      = error "copyProg: missing source parameter."
+copyProg opts Nothing _ = []
+copyProg opts (Just outExp) inExp =
   case inExp of
-    [x] | outExp == x -> Empty
-    _                 -> call "copy" (map ValueParameter (outExp:inExp))
+    [x] | outExp == x -> []
+    _                 -> deepCopy opts outExp inExp
 
 -- | Expand copying of aggregate data to copying of components
-deepCopy :: Options -> [ActualParameter ()] -> [Program ()]
-deepCopy opts [ValueParameter arg1, ValueParameter arg2]
+deepCopy :: Options -> Expression () -> [Expression ()] -> [Program ()]
```

Comment:
Can Options be removed as parameter to this function and all the "ask backendOpts" in compileProg disappear?

Reply:
Yes, I missed that simplification. It is only needed when lowering. Will do!